### PR TITLE
Make `@onflow/cadence-parser` support node.js (Hybrid ESM/CJS)

### PR DESCRIPTION
**Pull request by Pure-Peace**
After merging this, `@onflow/cadence-parser` can be used directly in node.js,
developers can easily use this library to develop some libraries (such as a library for generating contract schemas, similar to typechain).

And I tested its use in `tools/astexplorer` (browser), everything works fine.

```javascript
const {CadenceParser} = require('@onflow/cadence-parser');
const fs = require('fs');
const path = require('path');

(async () => {
  const parser = await CadenceParser.create(
    await fs.promises.readFile(path.join(__dirname, './node_modules/@onflow/cadence-parser/dist/cadence-parser.wasm'))
  );

  const ast = parser.parse(`
  pub contract HelloWorld {
    pub fun hello() {
      log("Hello, world!")
    }
  }
`);

  console.log(ast);
})();
```

## Description

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by turbolent** [comment]
Thank you for this contribution @Pure-Peace!

File: npm-packages/cadence-parser/tsconfig.json [hunk 1]
```diff
@@ -13,9 +13,20 @@
     "sourceMap": true,
     "skipLibCheck": true,
     "baseUrl": ".",
-    "lib": ["dom", "es2015", "es2016"],
+    "esModuleInterop": true,
+    "lib": [
+      "dom",
+      "es2015",
+      "es2016"
```
Comment (10thfloor) [inline npm-packages/cadence-parser/tsconfig.json:20]:
```suggestion
      "es6",
      "es7"
```

File: npm-packages/cadence-parser/package.json [hunk 2]
```diff
@@ -7,21 +7,36 @@
...
+      "require": "./dist/cjs/index.js",
+      "import": "./dist/esm/index.js",
+      "types": "./dist/types/index.d.ts"
+    }
+  },
   "scripts": {
-    "build": "tsc && GOARCH=wasm GOOS=js go build -o ./dist/cadence-parser.wasm ../../runtime/cmd/parse",
+    "build": "npm run build:types && npm run build:esm && npm run build:cjs && GOARCH=wasm GOOS=js go build -o ./dist/cadence-parser.wasm ../../runtime/cmd/parse",
+    "build:types": "tsc --emitDeclarationOnly --module system --outDir dist/types",
+    "build:esm": "tsc --module es2015 --target es6 --outDir dist/esm --declaration false",
```
Comment (10thfloor) [inline npm-packages/cadence-parser/package.json:23]:
```suggestion
    "build:esm": "tsc --module es6 --target es6 --outDir dist/esm --declaration false",
```

**Review by turbolent — APPROVED** [review]
Thank you! Ready to get merged once the feedback from @10thfloor was addressed

File: npm-packages/cadence-parser/src/go.ts [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+    const argv = offset;
+    argvPtrs.forEach((ptr) => {
+      this.mem.setUint32(offset, ptr, true);
+      this.mem.setUint32(offset + 4, 0, true);
+      offset += 8;
+    });
+
+    // The linker guarantees global data starts from at least wasmMinDataAddr.
+    // Keep in sync with cmd/link/internal/ld/data.go:wasmMinDataAddr.
+    const wasmMinDataAddr = 4096 + 4096;
```
Comment (SaveTheRbtz) [inline npm-packages/cadence-parser/src/go.ts:576]: this is different in 1.18 https://github.com/golang/go/blob/1d9d99b7ce279f2af928f79cbc5906d99f29bb67/src/cmd/link/internal/ld/data.go#L2459-L2463

Maybe we should first bump `wasm_exec.js` to the latest and then convert it to typescript?

Comment (Pure-Peace) [reply npm-packages/cadence-parser/src/go.ts:576]: Yes, I started trying to do this

Comment (turbolent) [reply npm-packages/cadence-parser/src/go.ts:576]: Good catch! Yes, unfortunately Go's WebAssembly runtime (`wasm_exec.js`) is version specific and we should bump it to 1.18. We can do this in a separate PR though

**Pull request merged**